src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt

```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
+
+/**
+ * Traverses types in `ty` and collects all items that unresolved in current context.
+ */
+fun RsElement.getTypeReferencesInfoFromTy(
+    ty: Ty,
+    subst: Substitution = emptySubstitution
+): TypeReferencesInfo = getTypeReferencesInfo(this, listOf(ty), subst, ::collectImportSubjectsFromTy)
+
+private fun <T> getTypeReferencesInfo(
+    context: RsElement,
+    elements: Collection<T>,
+    subst: Substitution,
+    collector: (T, Substitution, MutableSet<RsQualifiedNamedElement>) -> Unit
+): TypeReferencesInfo {
+    val result = hashSetOf<RsQualifiedNamedElement>()
+    elements.forEach { collector(it, subst, result) }
+    return processRawImportSubjects(context, result)
+}
+
+private fun collectImportSubjectsFromTypeReferences(
+    context: RsElement,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    context.accept(object : RsVisitor() {
+        override fun visitTypeReference(reference: RsTypeReference) =
+            collectImportSubjectsFromTy(reference.type, subst, result)
+
+        override fun visitElement(element: RsElement) =
+            element.acceptChildren(this)
+    })
 }
+
+private fun collectImportSubjectsFromTy(
+    ty: Ty,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    ty.substitute(subst).visitWith(object : TypeVisitor {
+        override fun visitTy(ty: Ty): Boolean {
+            when (ty) {
+                is TyAdt ->
+                    result.add(ty.item)
+                is TyAnon ->
+                    result.addAll(ty.traits.map { it.element })
+                is TyTraitObject ->
+                    result.add(ty.trait.element)
+                is TyProjection -> {
+                    result.add(ty.trait.element)
+                    result.add(ty.target)
+                }
+            }
+            return ty.superVisitWith(this)
+        }
+    })
+}
+
+fun RsElement.importElements(elements: Set<RsQualifiedNamedElement>) {
+    if (!RsCodeInsightSettings.getInstance().importOutOfScopeItems) return
+    val importContext = ImportContext.from(project, this)
+    for (element in elements) {
+        val name = element.name ?: continue
+        val candidates = AutoImportFix.getImportCandidates(importContext, name, name) {
+            !(it.item is RsMod || it.item is RsModDeclItem || it.item.parent is RsMembers)
+        }
+        val candidate = candidates.firstOrNull { it.qualifiedNamedItem.item in elements }
+        candidate?.import(this)
+    }
+}
+
+/**
+ * Takes `rawImportSubjects` and filters items that are unresolved in the current `context`.
+ * Then splits this items into two sets:
+ * - Items that can be imported
+ * - Items that can't be imported
+ */
+private fun processRawImportSubjects(
+    context: RsElement,
+    rawImportSubjects: Set<RsQualifiedNamedElement>
+): TypeReferencesInfo {
+    val importSubjects = hashMapOf<String, MutableSet<RsQualifiedNamedElement>>()
```

Comment:
Why do we need this map at all?
Can't we just process all elements from `rawImportSubjects` directly?